### PR TITLE
Multi-GPU support for `FaissIndex`

### PR DESCRIPTION
**Pull request by rentruewang**
Per #3716 , current implementation does not take into consideration that `faiss` can run on multiple GPUs.

In this commit, I provided multi-GPU support for `FaissIndex` by modifying the device management in `IndexableMixin.add_faiss_index` and `FaissIndex.load`.

Now users are able to pass in
1. a positive integer (as usual) to use 1 GPU
2. a negative integer `-1` to use all GPUs
3. a list of integers e.g. `[0, 1]` to run only on those GPUs
4. Of course, passing in nothing still runs on CPU.

This closes: #3716

### CODE REVIEW

**Comment by rentruewang** [comment]
Any love guys?

**Review by albertvillanova — CHANGES_REQUESTED** [review]
Thank you. Great job. Just a few comments/questions below...

File: src/datasets/search.py
Comment (rentruewang) [reply src/datasets/search.py:330]: Oh yeah, sorry I missed it. Thanks!

File: src/datasets/search.py
Comment (rentruewang) [reply src/datasets/search.py]: Fixed in the latest commit

File: src/datasets/search.py
Comment (rentruewang) [reply src/datasets/search.py]: Fixed in the latest commit.

**Comment by rentruewang** [comment]
@albertvillanova Hi, is there anything else you want to add to this PR?

**Review by lhoestq — APPROVED** [review]
It looks all good for me thanks ! @albertvillanova do you want to do a final review ?

Maybe we can wait for https://github.com/huggingface/datasets/pull/3690 to be merged first, since it changes the documentation format - or we would en up with many conflicts

File: src/datasets/search.py
Comment (rentruewang) [reply src/datasets/search.py:327]: @lhoestq Thanks for the feedback! I am a big fan of f-strings as well. I was just concerned about support for `python3.5`, is that not currently supported?

**Comment by rentruewang** [comment]
@albertvillanova Is there anything else you want me to add / modify?

**Review by albertvillanova — APPROVED** [review]
@rentruewang thank you again for your work!

I think this is OK.

Just, as @lhoestq suggested, maybe we should wait for other PR to be merged before this one, so that we avoid conflicts with the docstrings:
- #3690

**Comment by lhoestq** [comment]
Hi @rentruewang , we updated the documentation on `master`, could you merge `master` into your branch please ?

**Comment by rentruewang** [comment]
@lhoestq I've merge `huggingface/datasets/master` into this PR. Please review. Thanks! 🤗

**Pull request merged**
